pkg/runtime/reconciler.go

```diff
@@ -298,6 +302,70 @@ func (r *resourceReconciler) handleCacheError(
 	return r.HandleReconcileError(ctx, desired, latest, requeue.NeededAfter(err, roleARNNotAvailableRequeueDelay))
 }
 
+func (r *resourceReconciler) handleAdoption(
+	ctx context.Context,
+	rm acktypes.AWSResourceManager,
+	desired acktypes.AWSResource,
+	rlog acktypes.Logger,
+) (acktypes.AWSResource, error) {
+	// If the resource is being adopted by force, we need to access
+	// the required field passed by annotation and attempt a read.
+
+	rlog.Info("Adopting Resource")
+	extractedFields, err := ExtractAdoptionFields(desired)
+	if err != nil {
+		return desired, ackerr.NewTerminalError(err)
+	}
+	if len(extractedFields) == 0 {
+		// TODO(michaelhtm) Here we need to figure out if we want to have an
+		// error or not. should we consider accepting values from Spec?
+		// And then we can let the ReadOne figure out if we have missing
+		// required fields for a Read
+		return nil, fmt.Errorf("failed extracting fields from annotation")
+	}
+	resolved := desired.DeepCopy()
+	err = resolved.PopulateResourceFromAnnotation(extractedFields)
+	if err != nil {
+		return nil, err
+	}
+
+	rlog.Enter("rm.EnsureTags")
+	err = rm.EnsureTags(ctx, resolved, r.sc.GetMetadata())
+	rlog.Exit("rm.EnsureTags", err)
+	if err != nil {
+		return resolved, err
+	}
+	rlog.Enter("rm.ReadOne")
+	latest, err := rm.ReadOne(ctx, resolved)
+	if err != nil {
+		return latest, err
+	}
+
+	if err = r.setResourceManaged(ctx, rm, latest); err != nil {
+		return latest, err
+	}
+
+	// Ensure tags again after adding the finalizer and patching the
+	// resource. Patching desired resource omits the controller tags
+	// because they are not persisted in etcd. So we again ensure
+	// that tags are present before performing the create operation.
+	rlog.Enter("rm.EnsureTags")
+	err = rm.EnsureTags(ctx, latest, r.sc.GetMetadata())
+	rlog.Exit("rm.EnsureTags", err)
+	if err != nil {
+		return latest, err
+	}
```

Comment:
Does this mean that the first reconciliation omits tags? then we eventually add them?

Reply:
Do we need to call `EnsureTags` at L333?